app/src/main/java/com/hedvig/app/feature/offer/OfferViewModel.kt

```diff
@@ -93,279 +84,152 @@ abstract class OfferViewModel : ViewModel() {
-                                        offerResponse.offerResult,
-                                        offerResponse.loginStatus,
-                                        offerResponse.paymentMethods,
-                                        dataCollectionStatus,
-                                        dataCollectionResult.await(),
-                                        insuranceProviderDisplayName.await()
-                                    )
-                                )
-                            }
-                        }
-                }
-            }
-        }
-    }
+    override val viewState: StateFlow<ViewState> = _viewState
         .stateIn(
             scope = viewModelScope,
             started = SharingStarted.WhileSubscribed(5000),
             initialValue = ViewState.Loading,
         )
 
-    private suspend fun loadQuoteIds() {
-        getQuoteIdsUseCase.invoke(quoteCartId)
-            .map { it.ids }
-            .fold(
-                ifLeft = { offerAndLoginStatus.value = OfferAndLoginStatus.Error },
-                ifRight = {
-                    hAnalytics.screenViewOffer(it)
-                    quoteIds = it
-                }
-            )
+    init {
+        loginStatusService.isViewingOffer = shouldShowOnNextAppStart
+        loginStatusService.persistOfferIds(quoteCartId, quoteIds)
+
+        offerRepository.offerFlow(quoteIds)
+            .onEach { toViewState(it) }
+            .launchIn(viewModelScope)
+
+        viewModelScope.launch {
+            offerRepository.queryAndEmitOffer(quoteCartId, quoteIds)
+        }
     }
 
-    private suspend fun loadQuotes(quoteIds: List<String>) {
-        getQuotesUseCase.invoke(quoteIds, quoteCartId).collect { result ->
-            result.fold(
-                ifLeft = { offerAndLoginStatus.value = OfferAndLoginStatus.Error },
-                ifRight = { offerModel ->
-                    offerAndLoginStatus.value = OfferAndLoginStatus.Content(
-                        offerResult = offerModel,
+    private suspend fun toViewState(offerResult: Either<ErrorMessage, OfferModel>) = offerResult.fold(
+        ifLeft = { ViewState.Error(it.message) },
+        ifRight = { offerModel ->
+            getExternalInsuranceProviderUseCase
+                .observeExternalProviderOrNull(offerModel.externalProviderId)
+                .onEach {
+                    _viewState.value = ViewState.Content(
+                        offerModel = offerModel,
                         loginStatus = loginStatusService.getLoginStatus(),
-                        paymentMethods = getPaymentMethods()
+                        paymentMethods = offerModel.paymentApiResponseOrNull()
+                            ?: adyenRepository.paymentMethodsResponse(),
+                        externalProvider = it
                     )
                 }
-            )
+                .catch { _viewState.value = ViewState.Error(null) }
+                .launchIn(viewModelScope)
```

Comment:
Hmm might have to be careful with a function called `toViewState` which feels like a mapping of a model to a viewState representation which is launching a never-ending coroutine. Especially considering that if we call `toViewState` again a second time, `launchIn` is called a second time with nothing cancelling the old collection (since `launchIn` does `scope.launch{ collect() }` under the hood). 
Kind of hard to follow how this all would function in such scenarios. Again don't have a clear fix for this, and I see that you had to refactor a pretty big/complex function so I can see how some of the features it had can be hard to maintain in this process.
The previous implementation was fixing this by having the collection be inside `transformLatest` or `collectLatest` which meant that previous collecting jobs would be cancelled when new data is incoming, but can't see how to apply the same thing now, at least not from the GitHub code view :D Kinda hard to follow everything